### DISCUSSION TITLE
[Question][Hw3] isAssertFail and isAssume are missing / Which instructions should we support?

### DISCUSSION THREAD

**Discussion by park-sunho** [prosyslab-classroom/cs424-program-reasoning#320] [Q&A]
Hello, I have several questions on Hw3.

1. In README.md, it is written that skeleton code will provide `isAssertFail` and `isAssume`, but I cannot see both of them in my repo. Can I ask whether they are intentionally not provided or just missing?
2. I guess we should write some code for understanding the semantics of LLVM to create constraints in each CHC. Since there are lots of instructions that LLVM support and we've done similar work for limited set of instructions in Hw1, I guess this applies to Hw3 as well. In fact, I found some instructions such as `zext` that we did not support in Hw1 but we can see in this homework, so I think we should support more instructions than in Hw1. Can you let me know what kind of instructions we should support?

Thank you in advance.

**Comment by Hhro** [comment]
Hi,
Sorry for the confusion.

1. `isAssertFail` and `isAssume` are implemented in `llvmutils` as `is_assert_fail` and `is_assume`. You can use these APIs by calling `Llvmutils.is_assert_fail` and `Llvmutils.is_assume`.
2. For clarity, here are the operators you need to encode: 
    1. Binary operators: `add`, `sub`, `mul`, `udiv`, `sdiv`
    2. Terminator instructions: `br`
    3. Conversion operators: `zext`
    4. Other operators: `icmp`, `phi`, `call`

Thanks.

**Comment by Hhro** [comment]
I have updated the comment for question 2.
Those operators are all you need to encode for this homework.
Hope this helps.

**Comment by park-sunho** [comment]
Thank you so much.
